mongo/integration/unified/logger_verification.go

```diff
@@ -0,0 +1,283 @@
+type logMessage struct {
+	LevelLiteral      string   `bson:"level"`
+	ComponentLiteral  string   `bson:"component"`
+	Data              bson.Raw `bson:"data"`
+	FailureIsRedacted bool     `bson:"failureIsRedacted"`
+}
+
+// newLogMessage will create a "logMessage" from the level and a slice of
+// arguments.
+func newLogMessage(level int, args ...interface{}) (*logMessage, error) {
+	logMessage := new(logMessage)
+
+	// Iterate over the literal levels until we get the first
+	// "LevelLiteral" that matches the level of the "LogMessage". It doesn't
+	// matter which literal is chose so long as the mapping results in the
+	// correct level.
+	for literal, logLevel := range logger.LevelLiteralMap {
+		if level == int(logLevel) {
+			logMessage.LevelLiteral = literal
+
+			break
+		}
+	}
+
+	if len(args) == 0 {
+		return logMessage, nil
+	}
+
+	// The argument slice must have an even number of elements, otherwise it
+	// would not maintain the key-value structure of the document.
+	if len(args)%2 != 0 {
+		return nil, fmt.Errorf("invalid arguments: %v", args)
+	}
+
+	// Create a new document from the arguments.
+	actualD := bson.D{}
+	for i := 0; i < len(args); i += 2 {
+		actualD = append(actualD, bson.E{
+			Key:   args[i].(string),
+			Value: args[i+1],
+		})
+	}
+
+	// Marshal the document into a raw value and assign it to the
+	// logMessage.
+	bytes, err := bson.Marshal(actualD)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal: %v", err)
+	}
+
+	logMessage.Data = bson.Raw(bytes)
+
+	return logMessage, nil
+}
+
+// validate will validate the expectedLogMessage and return an error if it is
+// invalid.
+func validateLogMessage(message *logMessage) error {
+	if message.LevelLiteral == "" {
+		return fmt.Errorf("level is required")
+	}
+
+	if message.ComponentLiteral == "" {
+		return fmt.Errorf("component is required")
+	}
+
+	if message.Data == nil {
+		return fmt.Errorf("data is required")
+	}
+
+	return nil
+}
+
+// verifyLogMessagesMatch will verify that the actual log messages match the
+// expected log messages.
+func verifyLogMessagesMatch(ctx context.Context, exp, act *logMessage) error {
+	if act == nil && exp == nil {
+		return nil
+	}
+
+	if act == nil || exp == nil {
+		return errLogDocumentMismatch
+	}
+
+	levelExp := logger.ParseLevel(exp.LevelLiteral)
+	levelAct := logger.ParseLevel(act.LevelLiteral)
+
+	// The levels of the expected log message and the actual log message
+	// must match, upto logger.Level.
+	if levelExp != levelAct {
+		return fmt.Errorf("level mismatch: want %v, got %v", levelExp, levelAct)
+	}
+
+	rawExp := documentToRawValue(exp.Data)
+	rawAct := documentToRawValue(act.Data)
+
+	// Top level data does not have to be 1-1 with the expectation, there
+	// are a number of unrequired fields that may not be present on the
+	// expected document.
+	if err := verifyValuesMatch(ctx, rawExp, rawAct, true); err != nil {
+		return fmt.Errorf("%w: %v", errLogDocumentMismatch, err)
```

Comment:
Optional: Wrapping `errLogDocumentMismatch` doesn't provide much additional information because it won't have an error chain associated with it. Use the `%w` format verb for `err` instead. Also, it's a lot clearer to just write `"document mismatch: %w"` than to use an error variable.

```suggestion
		return fmt.Errorf("document mismatch: %w", err)
```